### ISSUE TITLE
BigQuery: KeyError inserting data with missing value into table with nullable field

### ISSUE THREAD

**Issue by jvalle3** [googleapis/google-cloud-python#7294]
I recently updated the google-cloud-bigquery python package and I am getting a **KeyError** when trying to insert a row with a **missing value** into a table with a **nullable field**.
In previous version (I tested it on version 1.5.0) the missing field was automatically filled with a _null_ value.

#### Environment details

OS: Ubuntu 18.04
Python version:
```
$ python --version
Python 3.6.7
```
Package version: 
```
$ pip show google-cloud-bigquery
Name: google-cloud-bigquery
Version: 1.9.0
```
Note: It works on version 1.5.0

#### Steps to reproduce

  1. Create table with nullable field
  2. Insert row with missing value (provide row as dictionary, with a missing key:value)

#### Code example

```python
from google.cloud import bigquery

client = bigquery.Client(project='my_project')

dataset_ref = client.dataset('my_dataset')
schema = [
    bigquery.SchemaField("user_id", "STRING", mode="REQUIRED"),
    bigquery.SchemaField("name", "STRING", mode="NULLABLE"), # NULLABLE
]
table_ref = dataset_ref.table("my_table")
table = bigquery.Table(table_ref, schema=schema)
table = client.create_table(table)

rows_to_insert = [
    {'user_id': 1111, 'name': 'peter'},  # OK
    {'user_id': 2222, 'name': 'anna'}  # OK
]

errors = client.insert_rows(table, rows_to_insert)
assert errors == []

rows_to_insert = [
    {'user_id': 3333, 'name': 'john'},
    {'user_id': 4444} # FAILS
]

errors = client.insert_rows(table, rows_to_insert) # FAILS HERE
assert errors == []
```

#### Stack trace
```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-22-a6f1bda68730> in <module>
     26 ]
     27 
---> 28 errors = client.insert_rows(table, rows_to_insert) # FAILS HERE
     29 assert errors == []

~/.local/lib/python3.6/site-packages/google/cloud/bigquery/client.py in insert_rows(self, table, rows, selected_fields, **kwargs)
   1495             raise TypeError("table should be Table or TableReference")
   1496 
-> 1497         json_rows = [_record_field_to_json(schema, row) for row in rows]
   1498 
   1499         return self.insert_rows_json(table, json_rows, **kwargs)

~/.local/lib/python3.6/site-packages/google/cloud/bigquery/client.py in <listcomp>(.0)
   1495             raise TypeError("table should be Table or TableReference")
   1496 
-> 1497         json_rows = [_record_field_to_json(schema, row) for row in rows]
   1498 
   1499         return self.insert_rows_json(table, json_rows, **kwargs)

~/.local/lib/python3.6/site-packages/google/cloud/bigquery/_helpers.py in _record_field_to_json(fields, row_value)
    399     for subindex, subfield in enumerate(fields):
    400         subname = subfield.name
--> 401         subvalue = row_value[subname] if isdict else row_value[subindex]
    402         record[subname] = _field_to_json(subfield, subvalue)
    403     return record

KeyError: 'name'
```

**Comment by tseaver** [comment]
/cc @adamf 

This change was introduced in e633ffdb96acebbd8c87d1f52edeb2432f6ac7e0, which fixed parsing nested schemas.

I think the change is as simple as using `row_value.get(subname)` instead of `row_value[subname]` in line 401.  Of course, we should add a test which exercises the missing value logic.

**Comment by kal1s** [comment]
Hey Guys, for information:

   I'm having the same problem but passing a list with dicts inside to be inserted in BigQuery Table. In my old Cloud Functions it's working fine yet but I needed to change other Cloud Function and start a lot of errors because is a pipeline of ETL from Cloud Storage to BigQuery. Locally worked fine until I had updated today google-cloud-bigquery to last version.

- Follow my version:
```
$ python3.7 -m pip show google-cloud-bigquery
Name: google-cloud-bigquery
Version: 1.9.0
```

- Error ocurred locally and on Cloud Functions today:

```
errorsInsertRows = gcpBQAccess.insert_rows(eventsTableBQ, eventsList)
  File "/home/kal1s/.local/lib/python3.7/site-packages/google/cloud/bigquery/client.py", line 1497, in insert_rows
    json_rows = [_record_field_to_json(schema, row) for row in rows]
  File "/home/kal1s/.local/lib/python3.7/site-packages/google/cloud/bigquery/client.py", line 1497, in <listcomp>
    json_rows = [_record_field_to_json(schema, row) for row in rows]
  File "/home/kal1s/.local/lib/python3.7/site-packages/google/cloud/bigquery/_helpers.py", line 401, in _record_field_to_json
    subvalue = row_value[subname] if isdict else row_value[subindex]
KeyError: 'smb_status_code'
```

The "smb_status_code" is a Key:Value like: {"smb_status_code": None} inside the list "eventsList".

Thank You!

**Comment by kal1s** [comment]
> Hey Guys, for information:
> 
> I'm having the same problem but passing a list with dicts inside to be inserted in BigQuery Table. In my old Cloud Functions it's working fine yet but I needed to change other Cloud Function and start a lot of errors because is a pipeline of ETL from Cloud Storage to BigQuery. Locally worked fine until I had updated today google-cloud-bigquery to last version.
> 
> * Follow my version:
> 
> ```
> $ python3.7 -m pip show google-cloud-bigquery
> Name: google-cloud-bigquery
> Version: 1.9.0
> ```
> * Error ocurred locally and on Cloud Functions today:
> 
> ```
> errorsInsertRows = gcpBQAccess.insert_rows(eventsTableBQ, eventsList)
>   File "~.local/lib/python3.7/site-packages/google/cloud/bigquery/client.py", line 1497, in insert_rows
>     json_rows = [_record_field_to_json(schema, row) for row in rows]
>   File "~.local/lib/python3.7/site-packages/google/cloud/bigquery/client.py", line 1497, in <listcomp>
>     json_rows = [_record_field_to_json(schema, row) for row in rows]
>   File "~.local/lib/python3.7/site-packages/google/cloud/bigquery/_helpers.py", line 401, in _record_field_to_json
>     subvalue = row_value[subname] if isdict else row_value[subindex]
> KeyError: 'smb_status_code'
> ```
> The "smb_status_code" is a Key:Value like: {"smb_status_code": None} inside the list "eventsList".
> 
> Thank You!


Important Additional Information:

   The files we are processing doesn't hit the smb_status_code, so when was working, BigQuery automatically placed "null" in this column and the other columns. We have 72 columns in the table and one row doesn't have data for all columns when we are inserting trough this way:

errorsInsertRows = gcpBQAccess.insert_rows(eventsTableBQ, eventsList)

Best Regards,

**Comment by kal1s** [comment]
> > Hey Guys, for information:
> > I'm having the same problem but passing a list with dicts inside to be inserted in BigQuery Table. In my old Cloud Functions it's working fine yet but I needed to change other Cloud Function and start a lot of errors because is a pipeline of ETL from Cloud Storage to BigQuery. Locally worked fine until I had updated today google-cloud-bigquery to last version.
> > 
> > * Follow my version:
> > 
> > ```
> > $ python3.7 -m pip show google-cloud-bigquery
> > Name: google-cloud-bigquery
> > Version: 1.9.0
> > ```
> > 
> > * Error ocurred locally and on Cloud Functions today:
> > 
> > ```
> > errorsInsertRows = gcpBQAccess.insert_rows(eventsTableBQ, eventsList)
> >   File "~.local/lib/python3.7/site-packages/google/cloud/bigquery/client.py", line 1497, in insert_rows
> >     json_rows = [_record_field_to_json(schema, row) for row in rows]
> >   File "~.local/lib/python3.7/site-packages/google/cloud/bigquery/client.py", line 1497, in <listcomp>
> >     json_rows = [_record_field_to_json(schema, row) for row in rows]
> >   File "~.local/lib/python3.7/site-packages/google/cloud/bigquery/_helpers.py", line 401, in _record_field_to_json
> >     subvalue = row_value[subname] if isdict else row_value[subindex]
> > KeyError: 'smb_status_code'
> > ```
> > The "smb_status_code" is a Key:Value like: {"smb_status_code": None} inside the list "eventsList".
> > Thank You!
> 
> Important Additional Information:
> 
> The files we are processing doesn't hit the smb_status_code, so when was working, BigQuery automatically placed "null" in this column and the other columns. We have 72 columns in the table and one row doesn't have data for all columns when we are inserting trough this way:
> 
> errorsInsertRows = gcpBQAccess.insert_rows(eventsTableBQ, eventsList)
> 
> Best Regards,

Hi Guys,
  
   We found another way to solve our problem, we changed from `insert_rows` to `insert_rows_json`.

   We tested locally and on Cloud Functions reprocessing 44k of files from Google Storage to BigQuery totalizing 700 MB without KeyErrors.

   We need to change because we insert rows with some values with "null".

   What worried us about it was since January 2 of this year `insert_rows` worked fine for us with the old version of google-cloud-bigquery.

   Now we will change other Cloud Functions to correct this call.

   We found that solution here: [google.cloud.bigquery.client.Client.insert_rows_json](https://google-cloud.readthedocs.io/en/latest/bigquery/generated/google.cloud.bigquery.client.Client.insert_rows_json.html)


Thanks!

**Comment by meirelesmj** [comment]
Hi @kal1s 

You just changed the function from `insert_rows` to `insert_rows_json' ?

Locally I am having no errors but with google cloud still have assertation errors.

You are passing a list of json tuples or a list of dicts ?

If you could share the code working it would be great.

Thanks
Manuel

**Comment by kal1s** [comment]
> Hi @kal1s
> 
> You just changed the function from `insert_rows` to `insert_rows_json' ?
> 
> Locally I am having no errors but with google cloud still have assertation errors.
> 
> You are passing a list of json tuples or a list of dicts ?
> 
> If you could share the code working it would be great.
> 
> Thanks
> Manuel

Hi @meirelesmj

   Our list has a lot of dicts inside and no one has all columns to be inserted in BigQuery, so BigQuery takes care of that column that doesn't have value putting "null" when we execute the **insert_rows_json**.

   Before I updated the Python 3.7 modules worked fine with **insert_rows** locally but not in Cloud Functions, so I decided to update my modules locally so stoped worked with same error on Cloud Functions. Looking around I founded it [google.cloud.bigquery.client.Client.insert_rows_json](https://google-cloud.readthedocs.io/en/latest/bigquery/generated/google.cloud.bigquery.client.Client.insert_rows_json.html), so I changed my code like I will explain next.

   Because Cloud Functions use Python 3.7 I'm using Python 3.7 locally too, so first I updated my Python 3.7 modules. I have installed locally the module google-cloud-bigquery, so in that update, this module will be updated too if is outdated.

**Update Python 3.7 modules:**

`python3.7 -m pip install -U $(python3.7 -m pip freeze | cut -d '=' -f 1)`


  After finished the updated, I changed my code.

**Update Code, from:**

`errorsInsertRows = gcpBQAccess.insert_rows(eventsTableBQ, eventsList)`

**Update Code, to:**

`errorsInsertRows = gcpBQAccess.insert_rows_json(eventsTableBQ, eventsList)`


   Now our batch and streaming are working normally.


I hope I've helped you...


Best Regards,

**Comment by meirelesmj** [comment]
Thanks @kal1s I am using 2.7 and its working ok with insert_rows_json but not in cloud functions.

Will change to python 3.7 locally to see if I have any errors like in Cloud Function.

In terms of code change on cloud functions was just that right ? no other imports needed right ?

Thanks for this great fix

**Comment by kal1s** [comment]
> Thanks @kal1s I am using 2.7 and its working ok with insert_rows_json but not in cloud functions.
> 
> Will change to python 3.7 locally to see if I have any errors like in Cloud Function.
> 
> In terms of code change on cloud functions was just that right ? no other imports needed right ?
> 
> Thanks for this great fix

Hi @meirelesmj 

   On your, Cloud Functions the `requirements.txt` need to have at least `google-cloud-bigquery` more other modules you are using in your function and change your code to `insert_rows_json` if you need to insert rows that are missing values on few columns where BigQuery will put "null" for you in your table.

   I recommend to you to use Python 3.7 locally too because are a lot of differences between Python 2 and 3, like coding and performance, and because Cloud Functions in this moment have support only to Python version 3.7 to execute our functions.

Best Regards,

**Comment by meirelesmj** [comment]
Hi @kal1s 

Can you paste your requirements. Are you setting the version for the library to 1.9.0


My current ones are

# Function dependencies, for example:
google-cloud
google-cloud-bigquery
google-cloud-storage
google-cloud-pubsub
xmltodict
dpath 
datetime


and in code

import json 
from google.cloud import bigquery
from google.cloud import storage 
import xmltodict 
import dpath.util

since in the requirements I am not forcing the version to 1.9.0 could it be using an older version ? 

Strange because now in local it runs with success but in the cloud function still the assertion errors


Thanks

**Comment by kal1s** [comment]
Hi @tseaver 

   I don't know if in this issue is the right place to discuss that, so please tell us it is not.

Hi @meirelesmj 

   I'm the concern if your Cloud Functions is Production if are please open a case on [Google Cloud Platform Support](https://cloud.google.com/support/) to they help you in more depth way because they have much more information about the Cloud Functions Environment.

   Follow my `requeriments.txt`

```
google-cloud-bigquery
google-cloud-storage
```

   Now, as we don't have access to the infrastructure of Cloud Functions, using coding we can find out the version of the module.

   So, in your code in Cloud Functions you need to use print after the import statement, like that:

```
import json
from google.cloud import bigquery
from google.cloud import storage
import xmltodict
import dpath.util

print("BigQuery VERSION:", bigquery.__version__)

print("Storage VERSION:", storage.__version__)
```

   You will need to look on your Cloud Functions Log for these prints. To help you, filter using `"BigQuery VERSION"` or `"Storage VERSION"`

   You can do the same thing locally to find out what version your Python Interpreter is using.

   In your locally code, look for the Python Version that your code is using adding this code:

```
import sys

print("Python VERSION:", sys.version)
```

   I don't know if `import sys` work on Cloud Functions, so you can try.

I hope I've helped you...

Best Regards,

**Comment by tseaver** [comment]
@kal1s The fix from #7297 is not yet released.

**Comment by kal1s** [comment]
Hi @tseaver 

   Now make sense why my code doesn't work with `insert_rows` even updating the Python Modules.

Thanks for the information!

Best Regards,
